service-api/src/test/kotlin/io/sweers/catchup/service/api/EmojiMarkdownConverterTest.kt

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) 2017 Zac Sweers
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.sweers.catchup.service.api
+
+import com.google.common.truth.Truth.assertThat
+import org.junit.Test
+
+class EmojiMarkdownConverterTest {
+
+  val replaced = "replaced"
+  val emoji = ":emoji:"
+  val converter = { alias: String ->  if (alias == "emoji") replaced else null }
+
+  @Test
+  fun testEmpty() {
+    val converted = replaceMarkdownEmojis("", converter)
+    assertThat(converted).isEmpty()
+  }
+
+  @Test
+  fun testSimpleReplace() {
+    val converted = convert(emoji)
+    assertThat(converted).isEqualTo(replaced)
+  }
+
+  @Test
+  fun testSimpleNoReplace() {
+    val converted = convert("emoji")
+    assertThat(converted).isEqualTo("emoji")
+  }
+
+  @Test
+  fun testReplaceOnceWithOtherText() {
+    val converted = convert("other text $emoji")
+    assertThat(converted).isEqualTo("other text $replaced")
+  }
+
+  @Test
+  fun testReplaceOnceWithExtraColons() {
+    var converted = convert(":$emoji")
+    assertThat(converted).isEqualTo(":$replaced")
+
+    converted = convert("$emoji:")
+    assertThat(converted).isEqualTo("$replaced:")
+
+    converted = convert(":$emoji:")
+    assertThat(converted).isEqualTo(":$replaced:")
+
+    converted = convert(":other text $emoji")
+    assertThat(converted).isEqualTo(":other text $replaced")
+  }
+
+  // Does this scenario matter, think we'd need to move away from `Regex(":(\\w+):")` to solve.
+  @Test
+  fun FAILING_ONE_REPLACE() {
```

Comment:
Note

Reply:
As in this test doesn't currently pass and we're wondering if it should?

Reply:
From what I can tell, this test won't pass with any `Regex`. The only solution I can think of is a simple string search that looks for `:` and see's if there's an emoji alias.

Left it in since I had already written the `Replacer` before realizing it fails on this scenario and wanted to get your opinion on the cleanest solution.

Reply:
One possible solution

```
fun replaceMarkdownEmojis(markdown: String, converter: EmojiMarkdownConverter): String {
  val sb = StringBuilder(markdown.length)
  var potentialAliasStart: Int? = null

  markdown.forEachIndexed { index, char ->
    if (char == ':') {
      potentialAliasStart = if (potentialAliasStart == null) {
        // If we have no potential start, any : is a potential start
        index
      } else {
        val potentialAlias = markdown.substring(potentialAliasStart!!, index + 1)
        val potentialEmoji = converter.convert(potentialAlias)
        // If we find an emoji append it and reset alias start, if we don't find an emoji
        // append between the potential start and this index *and* consider this index the new
        // potential start.
        if (potentialEmoji != null) {
          sb.append(potentialEmoji)
          null
        } else {
          sb.append(markdown, potentialAliasStart!!, index)
          index
        }
      }
      // While not looking for an alias end append all non possible alias chars to the string
    } else if (potentialAliasStart == null) {
      sb.append(char)
    }
  }

  // Finished iterating markdown while looking for an end, append anything remaining
  if (potentialAliasStart != null) {
    sb.append(markdown, potentialAliasStart!!, markdown.length)
  }

  return sb.toString()
}
```